test/lib/hierarchy-operations/hierarchy-operations.spec.js

```diff
@@ -0,0 +1,764 @@
+    });
+  });
+
+  it('documents should be minified', async () => {
+    await updateHierarchyRules([{ id: 'clinic', parents: ['district_hospital'] }]);
+    const patient = {
+      parent: parentsToLineage('clinic_1', 'health_center_1', 'district_1'),
+      type: 'person',
+      important: true,
+    };
+    const clinic = {
+      parent: parentsToLineage('health_center_1', 'district_1'),
+      type: 'clinic',
+      important: true,
+    };
+    patient.parent.important = false;
+    clinic.parent.parent.important = false;
+
+    await upsert('clinic_1', clinic);
+    await upsert('patient_1', patient);
+
+    await HierarchyOperations(pouchDb).move(['clinic_1'], 'district_2');
+
+    expect(getWrittenDoc('clinic_1')).to.deep.eq({
+      _id: 'clinic_1',
+      type: 'clinic',
+      important: true,
+      parent: parentsToLineage('district_2'),
+    });
+    expect(getWrittenDoc('patient_1')).to.deep.eq({
+      _id: 'patient_1',
+      type: 'person',
+      important: true,
+      parent: parentsToLineage('clinic_1', 'district_2'),
+    });
+  });
+
+  it('cannot create circular hierarchy', async () => {
+    // even if the hierarchy rules allow it
+    await updateHierarchyRules([{ id: 'health_center', parents: ['clinic'] }]);
+
+    try {
+      await HierarchyOperations(pouchDb).move(['health_center_1'], 'clinic_1');
+      assert.fail('should throw');
+    } catch (err) {
+      expect(err.message).to.include('circular');
+    }
+  });
+
+  it('throw if parent does not exist', async () => {
+    const actual = HierarchyOperations(pouchDb).move(['clinic_1'], 'dne_parent_id');
+    await expect(actual).to.eventually.rejectedWith('could not be found');
+  });
+
+  it('throw when altering same lineage', async () => {
+    const actual = HierarchyOperations(pouchDb).move(['patient_1', 'health_center_1'], 'district_2');
+    await expect(actual).to.eventually.rejectedWith('same lineage');
+  });
+
+  it('throw if contact_id is not a contact', async () => {
+    const actual = HierarchyOperations(pouchDb).move(['report_1'], 'clinic_1');
+    await expect(actual).to.eventually.rejectedWith('unknown type');
+  });
+
+  it('throw if moving primary contact of parent', async () => {
+    const actual = HierarchyOperations(pouchDb).move(['clinic_1_contact'], 'district_1');
+    await expect(actual).to.eventually.rejectedWith('primary contact');
+  });
+
+  it('throw if setting parent to self', async () => {
+    await updateHierarchyRules([{ id: 'clinic', parents: ['clinic'] }]);
+    const actual = HierarchyOperations(pouchDb).move(['clinic_1'], 'clinic_1');
+    await expect(actual).to.eventually.rejectedWith('circular');
+  });
+
+  it('throw when moving place to unconfigured parent', async () => {
+    await updateHierarchyRules([{ id: 'district_hospital', parents: [] }]);
+    const actual = HierarchyOperations(pouchDb).move(['district_1'], 'district_2');
+    await expect(actual).to.eventually.rejectedWith('parent of type');
+  });
+
```

Comment:
suggestion: can we add a few more error cases:

```suggestion

  it('throw if source does not exist', async () => {
    const nonExistentId = 'dne_parent_id';
    const actual = HierarchyOperations(pouchDb).move(['health_center_1',  nonExistentId], 'district_2');
    await expect(actual).to.eventually.rejectedWith(`Contact with id '${nonExistentId}' could not be found.`);
  });

  it('throw if ancestor does not exist', async () => {
    const sourceId = 'health_center_1';
    const healthCenter = await pouchDb.get(sourceId);
    healthCenter.name = 'no parent';
    healthCenter.parent._id = 'dne_parent_id';
    await pouchDb.put(healthCenter);

    const actual = HierarchyOperations(pouchDb).move([sourceId], 'district_2');
    await expect(actual).to.eventually.rejectedWith(
      `Contact '${healthCenter.name}' (${sourceId}) has parent id(s) '${healthCenter.parent._id}' which could not be found.`
    );
  });

```